runtime/compiler/optimizer/InterpreterEmulator.hpp

```diff
@@ -0,0 +1,253 @@
+      TR_ALLOC(TR_Memory::Inliner);
+      virtual IconstOperand *asIconst(){ return NULL;}
+      virtual KnownObjOperand *asKnownObject(){ return NULL;}
+      virtual KnownObjPairOperand* asKnownObjectPair(){ return NULL;}
+      virtual bool isUnkownOperand(){ return true;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return TR::KnownObjectTable::UNKNOWN;}
+   };
+
+/*
+ * \class KnownOperand
+ *
+ * \brief represent operands that can be reasoned about at compile time
+ */
+class KnownOperand : public Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      virtual bool isUnkownOperand(){ return false;}
+   };
+
+class IconstOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      IconstOperand (int x): intValue(x) { }
+      virtual IconstOperand *asIconst() { return this;}
+      int32_t intValue;
+   };
+
+class KnownObjOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      KnownObjOperand(TR::KnownObjectTable::Index koi):knownObjIndex(koi){ }
+      virtual KnownObjOperand *asKnownObject(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndex;}
+      TR::KnownObjectTable::Index knownObjIndex;
+   };
+
+/*
+ * \class KnownObjPairOperand
+ *
+ * \brief represents a pair of known objects
+ *
+ * \note This class is used to support mutable callsite because both the methodhandle object
+ *       and the mutable callsite needs to be tracked so that when creating \c TR_J9MutableCallSite
+ *       the mutable callsite object can be set for the callsite even though it's really the
+ *       methodhandle object that's on the operand stack.
+ *
+ * \see getReturnValueForInvokevirtual
+ * \see refineResolvedCalleeForInvokestatic
+ * \see visitInvokestatic
+ */
+class KnownObjPairOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      KnownObjPairOperand (TR::KnownObjectTable::Index koi1, TR::KnownObjectTable::Index koi2){ knownObjIndexPair = std::make_pair(koi1, koi2);}
+      virtual KnownObjPairOperand* asKnownObjectPair(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndexPair.first;}
+      std::pair<TR::KnownObjectTable::Index, TR::KnownObjectTable::Index> knownObjIndexPair;
+   };
+
+class InterpreterEmulator : public TR_ByteCodeIteratorWithState<TR_J9ByteCode, J9BCunknown, TR_J9ByteCodeIterator, Operand *>
+   {
+   typedef TR_ByteCodeIteratorWithState<TR_J9ByteCode, J9BCunknown, TR_J9ByteCodeIterator, Operand *> Base;
+
+   public:
+      InterpreterEmulator(
+            TR_CallTarget *calltarget,
+            TR::ResolvedMethodSymbol * methodSymbol,
+            TR_J9VMBase * fe,
+            TR::Compilation * comp,
+            TR_InlinerBase *inliner,
+            TR_EstimateCodeSize *ecs)
+         : Base(methodSymbol, comp),
+           _calltarget(calltarget),
+           _inliner(inliner),
```

Comment:
So looking at the code `_inliner` is used for two things - controlling tracing through `_inliner->tracer()` and accessing comp. I don't think there is a reason this abstract interpreter has to be tightly coupled to the inliner in that this class basically interprets bytecode, builds CallSite objects and calls a function to do something with that update - namely `findTargetAndUpdateInfoForCallsite`. That function could do almost anything if we wanted to re-purpose this in a new context.  Avoiding the tight coupling would make sense to me - thoughts?

Reply:
Modified the code to be less coupled to the inliner object. The InterpreterEmulator is having its own tracer object now rather than needing to get it through an inliner object. Currently, the tracer is still `InlinerTracer` type because of the way `TR_PrexArgInfo` does tracing.

Reply:
Ok I'm fine with this now - can we create an item to either factor that tracer out of the inliner or convert to a more conventional tracing mechanism so that this code is more reusable? It is not a must do for this item, but we should remember to come back. Could even be a more beginner item for someone...

Reply:
https://github.com/eclipse/openj9/issues/7936 is created to track this problem.